File: semanticsmith-py/src/semanticsmith/align.py
"""Main alignment API for semanticsmith."""

from __future__ import annotations

from typing import Callable, Sequence

import numpy as np

from semanticsmith.types import AlignmentResult, SimilarityFn

try:
    from semanticsmith._core import c_align
except ImportError:
    from semanticsmith._fallback import c_align

GAP_TOKEN = "---"


def _tokenize(text: str | list[str]) -> list[str]:
    """Convert input to a list of tokens."""
    if isinstance(text, str):
        return text.split()
    return list(text)


def _build_sim_matrix(
    tokens1: list[str],
    tokens2: list[str],
    similarity_fn: SimilarityFn,
) -> np.ndarray:
    """Build the n x m similarity matrix by calling the user's function."""
    n, m = len(tokens1), len(tokens2)
    matrix = np.empty((n, m), dtype=np.float64)
    for i in range(n):
        for j in range(m):
            matrix[i, j] = similarity_fn(tokens1[i], tokens2[j])
    return matrix


def align(
    text1: str | list[str],
    text2: str | list[str],
    similarity_fn: SimilarityFn,
    gap_open: float = -5.0,
    gap_extend: float = -2.0,
) -> AlignmentResult:
    """Align two texts using Semantic Smith-Waterman.

    Parameters
    ----------
    text1 : str or list of str
        First text (string to be tokenized by whitespace, or pre-tokenized list).
    text2 : str or list of str
        Second text.
    similarity_fn : callable (str, str) -> float
        Returns the similarity score between two words.
    gap_open : float
        Penalty for opening a gap (should be negative).
    gap_extend : float
        Penalty for extending a gap (should be negative).

    Returns
    -------
    AlignmentResult
    """
    tokens1 = _tokenize(text1)
    tokens2 = _tokenize(text2)

    if not tokens1 or not tokens2:
        return AlignmentResult(
            score=0.0, aligned1=[], aligned2=[],
            identity=0.0, n_gaps=0,
            seq1_tokens=tokens1, seq2_tokens=tokens2,
        )

    sim_matrix = _build_sim_matrix(tokens1, tokens2, similarity_fn)
    score, idx1, idx2 = c_align(sim_matrix, gap_open, gap_extend)

    aligned1 = []
    aligned2 = []
    n_gaps = 0
    matches = 0

    for i1, i2 in zip(idx1, idx2):
        t1 = tokens1[i1] if i1 >= 0 else GAP_TOKEN
        t2 = tokens2[i2] if i2 >= 0 else GAP_TOKEN
        aligned1.append(t1)
        aligned2.append(t2)
        if i1 < 0 or i2 < 0:
            n_gaps += 1
        elif tokens1[i1] == tokens2[i2]:
            matches += 1

    align_len = len(aligned1)
    identity = matches / align_len if align_len > 0 else 0.0

    return AlignmentResult(
        score=score,
        aligned1=aligned1,
        aligned2=aligned2,
        identity=identity,
        n_gaps=n_gaps,
        seq1_tokens=tokens1,
        seq2_tokens=tokens2,
    )


def batch_align(
    texts: Sequence[str | list[str]],
    similarity_fn: SimilarityFn,
    gap_open: float = -5.0,
    gap_extend: float = -2.0,
    pairs: Sequence[tuple[int, int]] | None = None,
) -> list[dict]:
    """Align multiple texts pairwise.

    Parameters
    ----------
    texts : sequence of str or list[str]
        Texts to compare.
    similarity_fn : callable (str, str) -> float
        Word similarity function.
    gap_open, gap_extend : float
        Gap penalties.
    pairs : sequence of (i, j) tuples, optional
        Specific pairs to compare. If None, computes all upper-triangle pairs.

    Returns
    -------
    list of dict with keys: i, j, score, identity, align_len, n_gaps, alignment
    """
    if pairs is None:
        n = len(texts)
        pairs = [(i, j) for i in range(n) for j in range(i + 1, n)]

    results = []
    for i, j in pairs:
        r = align(texts[i], texts[j], similarity_fn, gap_open, gap_extend)
        results.append({
            "i": i,
            "j": j,
            "score": r.score,
            "identity": r.identity,
            "align_len": r.align_len,
            "n_gaps": r.n_gaps,
            "alignment": r,
        })

    return results

File: r/R/align.R
#' Semantic Smith-Waterman Local Alignment
#'
#' Align two tokenized texts using the Smith-Waterman algorithm with
#' an arbitrary word similarity function and affine gap penalties.
#'
#' @param tokens1 Character vector of tokens, or a single string (split by whitespace).
#' @param tokens2 Character vector of tokens, or a single string.
#' @param sim_fn A function taking two strings and returning a numeric similarity score.
#' @param gap_open Penalty for opening a gap (should be negative). Default -5.
#' @param gap_extend Penalty for extending a gap (should be negative). Default -2.
#' @return A tibble with one row containing: score, identity, align_len, n_gaps,
#'   aligned1 (list-col), aligned2 (list-col).
#' @export
#' @examples
#' sim <- sim_from_function(function(a, b) ifelse(a == b, 10, -5))
#' ssw_align("the cat sat", "the big cat sat", sim)
ssw_align <- function(tokens1, tokens2, sim_fn,
                      gap_open = -5, gap_extend = -2) {
  tokens1 <- .tokenize(tokens1)
  tokens2 <- .tokenize(tokens2)

  n <- length(tokens1)
  m <- length(tokens2)

  if (n == 0 || m == 0) {
    return(tibble::tibble(
      score = 0, identity = 0, align_len = 0L, n_gaps = 0L,
      aligned1 = list(character(0)),
      aligned2 = list(character(0))
    ))
  }

  # Build similarity matrix
  sim_matrix <- matrix(0, nrow = n, ncol = m)
  for (i in seq_len(n)) {
    for (j in seq_len(m)) {
      sim_matrix[i, j] <- sim_fn(tokens1[i], tokens2[j])
    }
  }

  result <- c_ssw_align(sim_matrix, gap_open, gap_extend)

  # Convert indices to tokens (0 = gap in R-side result)
  aligned1 <- ifelse(result$align1 > 0, tokens1[result$align1], "---")
  aligned2 <- ifelse(result$align2 > 0, tokens2[result$align2], "---")

  n_gaps <- sum(result$align1 == 0 | result$align2 == 0)
  align_len <- length(aligned1)
  matches <- sum(aligned1 == aligned2 & aligned1 != "---")
  identity <- if (align_len > 0) matches / align_len else 0

  tibble::tibble(
    score = result$score,
    identity = identity,
    align_len = align_len,
    n_gaps = n_gaps,
    aligned1 = list(aligned1),
    aligned2 = list(aligned2)
  )
}


#' Batch Pairwise Alignment
#'
#' Align pairs of texts from a data frame. Pipe-friendly with tidyverse
#' conventions: data-first argument, tidy evaluation for column selection.
#'
#' @param .data A data frame or tibble.
#' @param col1 Column containing first texts (tidy-select, unquoted).
#' @param col2 Column containing second texts (tidy-select, unquoted).
#' @param sim_fn A similarity function (see [sim_from_function()]).
#' @param gap_open Gap opening penalty. Default -5.
#' @param gap_extend Gap extension penalty. Default -2.
#' @return The input tibble with added columns: ssw_score, ssw_identity,
#'   ssw_align_len, ssw_n_gaps, ssw_aligned1, ssw_aligned2.
#' @export
#' @examples
#' df <- tibble::tibble(
#'   text_a = c("the cat sat", "hello world"),
#'   text_b = c("the dog sat", "goodbye world")
#' )
#' sim <- sim_from_function(function(a, b) ifelse(a == b, 10, -5))
#' df |> ssw_batch_align(text_a, text_b, sim)
ssw_batch_align <- function(.data, col1, col2, sim_fn,
                            gap_open = -5, gap_extend = -2) {
  col1_name <- rlang::as_name(rlang::enquo(col1))
  col2_name <- rlang::as_name(rlang::enquo(col2))

  texts1 <- .data[[col1_name]]
  texts2 <- .data[[col2_name]]

  results <- lapply(seq_along(texts1), function(i) {
    ssw_align(texts1[[i]], texts2[[i]], sim_fn, gap_open, gap_extend)
  })

  result_tbl <- dplyr::bind_rows(results)

  .data$ssw_score <- result_tbl$score
  .data$ssw_identity <- result_tbl$identity
  .data$ssw_align_len <- result_tbl$align_len
  .data$ssw_n_gaps <- result_tbl$n_gaps
  .data$ssw_aligned1 <- result_tbl$aligned1
  .data$ssw_aligned2 <- result_tbl$aligned2

  .data
}


# Internal: split string into tokens if needed
.tokenize <- function(x) {
  if (length(x) == 1 && !is.na(x)) {
    tokens <- strsplit(x, "\\s+")[[1]]
    tokens[nchar(tokens) > 0]
  } else {
    as.character(x)
  }
}
